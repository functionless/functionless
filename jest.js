const swcJest = require("@swc/jest");
const { config } = require("./swc-config");

function createTransformer() {
  return swcJest.createTransformer(config);
}

module.exports = { createTransformer };
