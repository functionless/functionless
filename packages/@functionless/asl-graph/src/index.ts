export * from "./asl-graph";
export * from "./asl-integration";
export * from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export * from "./states";
export * from "./step-function-error";
