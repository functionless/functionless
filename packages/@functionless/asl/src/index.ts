export { ASLGraph } from "./asl-graph";
export { ASLIntegration, isASLIntegration } from "./asl-integration";
export { ASL } from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export {
  Branch,
  Catch,
  Choice,
  CommonFields,
  CommonTaskFields,
  Condition,
  Fail,
  MapTask,
  ParallelTask,
  Parameters,
  Pass,
  Retry,
  State,
  StateMachine,
  States,
  Succeed,
  Task,
  TerminalState,
  Wait,
  isChoiceState,
  isFailState,
  isMapTaskState,
  isParallelTaskState,
  isPassState,
  isState,
  isSucceedState,
  isTaskState,
  isWaitState,
} from "./states";
export { StepFunctionCause, StepFunctionError } from "./step-function-error";
