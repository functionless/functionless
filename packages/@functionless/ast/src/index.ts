export * from "./assert";
export * from "./constants";
export * from "./declaration";
export * from "./ensure";
export * from "./error";
export * from "./expression";
export * from "./find";
export * from "./guards";
export * from "./literal";
export * from "./node-clone";
export * from "./node-ctor";
export * from "./node-kind";
export * from "./node";
export * from "./reflect";
export * from "./resolve-references";
export * from "./s-expression";
export * from "./span";
export * from "./statement";
export * from "./util";
export * from "./visit";
