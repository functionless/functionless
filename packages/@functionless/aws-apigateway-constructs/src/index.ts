export { APIGatewayVTL } from "./api-vtl";
export {
  APIGatewayError,
  APIGatewayRequestOverride,
  ApiBody,
  ApiGatewayContext,
  ApiGatewayInput,
  ApiGatewayResponseOverride,
  ApiMethod,
  ApiMethodKind,
  ApiParameter,
  ApiParameters,
  ApiRequest,
  AwsMethod,
  HttpMethod,
  LambdaMethod,
  LambdaMethodProps,
  MethodProps,
  MockMethod,
  isApiMethodKind,
} from "./api";
