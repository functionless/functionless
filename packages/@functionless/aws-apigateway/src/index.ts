export * from "./api-vtl";
export * from "./api-integration";
