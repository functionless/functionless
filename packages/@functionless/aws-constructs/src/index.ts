export * from "../../aws-appsync-constructs/src/api";
export * from "./appsync";
export * from "./async-synth";
export * from "./aws";
export * from "./event-bridge";
export * from "./event-source";
export * from "./iterable";
export * from "./queue";
export * from "./queue";
export * from "./secret";
export * from "@functionless/aws-stepfunctions-constructs/src/step-function";
export * from "./table";
export * from "./user-pool";
export * from "./util";

// to prevent the closure serializer from trying to import all of functionless.
export const deploymentOnlyModule = true;
