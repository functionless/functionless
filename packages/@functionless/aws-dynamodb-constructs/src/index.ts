export * from "./appsync";
export * from "./batch-get-item";
export * from "./batch-write-item";
export * from "./delete-item";
export * from "./get-item";
export * from "./integration";
export * from "./put-item";
export * from "./query";
export * from "./runtime";
export * from "./scan";
export * from "./table";
export * from "./transact-get-item";
export * from "./transact-write-item";
export * from "./update-item";
export * from "./util";
