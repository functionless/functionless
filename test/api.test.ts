import "jest";
import { aws_apigateway, IResolvable, Stack } from "aws-cdk-lib";
import { AwsApiIntegration, MockApiIntegration, Function } from "../src";

let stack: Stack;
let func: Function<any, any>;
beforeEach(() => {
  stack = new Stack();
  func = new Function(stack, "F", (p) => p);
});

test("mock integration with object literal", () => {
  const api = new aws_apigateway.RestApi(stack, "API");

  const method = getMethodTemplates(
    new MockApiIntegration({
      request: (req: {
        pathParameters: {
          code: number;
        };
      }) => ({
        statusCode: req.pathParameters.code,
      }),
      responses: {
        200: () => ({
          response: "OK",
        }),
        500: () => ({
          response: "BAD",
        }),
      },
    }).addMethod("GET", api.root)
  );

  expect(method.httpMethod).toEqual("GET");
  expect(method.integration.requestTemplates).toEqual({
    "application/json": `#set($inputRoot = $input.path('$'))
{"statusCode":$input.params().path.code}`,
  });
  expect(method.integration.integrationResponses).toEqual([
    <IntegrationResponseProperty>{
      statusCode: "200",
      selectionPattern: "^200$",
      responseTemplates: {
        "application/json": `#set($inputRoot = $input.path('$'))
{"response":"OK"}`,
      },
    },
    <IntegrationResponseProperty>{
      statusCode: "500",
      selectionPattern: "^500$",
      responseTemplates: {
        "application/json": `#set($inputRoot = $input.path('$'))
{"response":"BAD"}`,
      },
    },
  ]);
});

test("AWS integration with Function", () => {
  const api = new aws_apigateway.RestApi(stack, "API");

  const method = getMethodTemplates(
    new AwsApiIntegration({
      request: (req: {
        pathParameters: {
          code: number;
        };
      }) => func(req),
      response: (result) => ({
        result,
      }),
    }).addMethod("GET", api.root)
  );

  expect(method.httpMethod).toEqual("GET");
  expect(method.integration.requestTemplates).toEqual({
    "application/json": `#set($inputRoot = $input.path('$'))
"$inputRoot"`,
  });
  expect(method.integration.integrationResponses).toEqual([
    <IntegrationResponseProperty>{
      statusCode: "200",
      responseTemplates: {
        "application/json": `#set($inputRoot = $input.path('$'))
{"result":"$inputRoot"}`,
      },
    },
  ]);
});

type CfnIntegration = Exclude<
  aws_apigateway.CfnMethod["integration"],
  IResolvable | undefined
> & {
  integrationResponses: IntegrationResponseProperty[];
};

interface IntegrationResponseProperty {
  readonly contentHandling?: string;
  readonly responseParameters?: {
    [key: string]: string;
  };
  readonly responseTemplates?: {
    [key: string]: string;
  };
  readonly selectionPattern?: string;
  readonly statusCode: string;
}

function getMethodTemplates(
  method: aws_apigateway.Method
): aws_apigateway.CfnMethod & {
  integration: CfnIntegration;
} {
  return method.node.findChild("Resource") as any;
}
