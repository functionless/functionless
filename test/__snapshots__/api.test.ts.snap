// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AWS integration with DynamoDB Table 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "#set($v1 = $input.path('$').Item != $null)
#if($v1)
{\\"data\\":$input.json('$.Item')}
#else
#set($context.responseOverride.status = 404)
$util.qr($context.responseOverride.status)
{\\"requestId\\":$input.json('$.requestId'),\\"missing\\":true}
#end",
      },
      "statusCode": "200",
    },
  ],
  "requestTemplates": Object {
    "application/json": "{
  \\"TableName\\":\\"__REPLACED_TOKEN\\",
  \\"Key\\":{\\"pk\\":{\\"S\\":$input.json('$.id')}}
}",
  },
}
`;

exports[`AWS integration with Express Step Function 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "#set($v1 = $input.path('$').status == 'SUCCEEDED')
#if($v1)
$input.json('$.output')
#else
#set($context.responseOverride.status = 500)
$util.qr($context.responseOverride.status)
$input.json('$.error')
#end",
      },
      "statusCode": "200",
    },
  ],
  "requestTemplates": Object {
    "application/json": "{
\\"stateMachineArn\\":\\"__REPLACED_TOKEN\\",
\\"input\\":\\"{\\\\\\"num\\\\\\":$input.params('num'),\\\\\\"str\\\\\\":\\\\\\"$input.params('str')\\\\\\"}\\"
}",
  },
}
`;

exports[`AWS integration with Function 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "{\\"result\\":$input.json('$')}",
      },
      "statusCode": "200",
    },
  ],
  "requestTemplates": Object {
    "application/json": "$input.json('$.prop')",
  },
}
`;

exports[`mock integration with object literal 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "{\\"response\\":\\"OK\\"}",
      },
      "selectionPattern": "^200$",
      "statusCode": "200",
    },
    Object {
      "responseTemplates": Object {
        "application/json": "{\\"response\\":\\"BAD\\"}",
      },
      "selectionPattern": "^500$",
      "statusCode": "500",
    },
  ],
  "requestTemplates": Object {
    "application/json": "{\\"statusCode\\":$input.params('code')}",
  },
}
`;

exports[`mock integration with object literal and literal type in pathParameters 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "{\\"response\\":\\"OK\\"}",
      },
      "selectionPattern": "^200$",
      "statusCode": "200",
    },
    Object {
      "responseTemplates": Object {
        "application/json": "{\\"response\\":\\"BAD\\"}",
      },
      "selectionPattern": "^500$",
      "statusCode": "500",
    },
  ],
  "requestTemplates": Object {
    "application/json": "{\\"statusCode\\":$input.params('code')}",
  },
}
`;

exports[`return $input.data 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "$input.json('$')",
      },
      "statusCode": "200",
    },
  ],
  "requestTemplates": Object {
    "application/json": "{
  \\"TableName\\":\\"__REPLACED_TOKEN\\",
  \\"Key\\":{\\"pk\\":{\\"S\\":$input.json('$.id')}}
}",
  },
}
`;

exports[`return $input.data.list[0] 1`] = `
Object {
  "integrationResponses": Array [
    Object {
      "responseTemplates": Object {
        "application/json": "$input.json('$')",
      },
      "statusCode": "200",
    },
  ],
  "requestTemplates": Object {
    "application/json": "{
  \\"TableName\\":\\"__REPLACED_TOKEN\\",
  \\"Key\\":{\\"pk\\":{\\"S\\":$input.json('$.list[0]')}}
}",
  },
}
`;
